File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    public int solution(int n, int[][] wires) {
        // 그래프 초기화
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int i = 1; i <= n; i++){
            graph.put(i, new ArrayList<>());
        }
        
        // 그래프 입력
        for (int[] wire: wires){
            int a = wire[0];
            int b = wire[1];
            
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        
        // 하나씩 끊어보기
        int answer = Integer.MAX_VALUE;
        for (int[]wire: wires){
            int a= wire[0];
            int b = wire[1];
            int count1 = bfs(a, graph, new int[]{a,b}, n);
            int count2 = n - count1;
            int diff = Math.abs(count1 - count2);
            answer = Math.min(answer,diff);
        }
        
        return answer;
    }
    
    int bfs(int start, Map<Integer, List<Integer>> graph, int[] cuttingEdge, int n){
        Deque <Integer> queue = new ArrayDeque<>();
        boolean[] visited = new boolean[n + 1];
        int count = 1;
        
        queue.offerLast(start);
        visited[start] = true;
        
        while (!queue.isEmpty()){
            int curr = queue.pollFirst();
            
            for (int next : graph.get(curr)){
                if (!visited[next] && 
                    !(Arrays.equals(cuttingEdge, new int[]{curr, next}) || Arrays.equals(cuttingEdge, new int[]{next, curr}))) {
                    visited[next] = true;
                    queue.offerLast(next);
                    count++;
                }
            }
        }
        return count;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque
def solution(n, wires):
    answer = int(1e9)
    ## 모든 전선을 끊어봐서 diff 가 가장 작은 경우의 수를 찾아야됨
    graph = {i : [] for i in range(1,n + 1)}
    for [a,b] in wires:
        graph[a].append(b)
        graph[b].append(a)
        
    def bfs(start, cutting_edge):
        queue = deque()
        visited = {i: False for i in range(1, n + 1)}
        queue.append(start)
        visited[start] = True
        count = 1
        
        while queue:
            current = queue.popleft()
            
            for next in graph[current]:
                if not visited[next] and not ([current,next] == cutting_edge or [next, current] == cutting_edge):
                    visited[next] = True
                    queue.append(next)
                    count += 1
        return count
    
    # 하나씩 다 끊어보기
    for [a,b] in wires:
        count1 = bfs(a, [a,b])
        count2 = n - count1
        diff = abs(count1 - count2)
        answer = min(answer, diff)
    return answer